### PR TITLE
2.13: move to Akka 2.6, Play 2.8, Lagom 1.6

### LINKED ISSUE #963: 2.13: move to Akka 2.6, revive Lagom
Lagom's 2.13 support is on their master branch where they have also moved to Akka 2.6. the CB is still on Akka 2.5

(it's true that we had Lagom on 2.13 green here for a while, but that was using a special branch, I don't see any value in continuing with that)

at some point we could/should move everything forward to Akka 2.6, but I think attempting to do so now would likely not have good cost/benefit. maybe once 2.6 is in RCs would be good timing? I could coordinate with the Akka and Play/Lagom teams

(likely the 2.12 build should stay on Akka 2.5 and Lagom 1.5 forever)

fyi @patriknw @dwijnand

### PR DESCRIPTION
**Pull request by SethTisue**
fixes #963

### CODE REVIEW

**Comment by SethTisue** [comment]
queued: https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2756/

File: proj/lagom.conf [hunk 1]
```diff
@@ -1,12 +1,14 @@
-// https://github.com/lagom/lagom.git#1.5.x
+// https://github.com/lagom/lagom.git#master
+
+// perhaps eventually there will be a 1.6.x branch for us to track;
+// at present (November 2019) 1.6.x development is happening on master
```
Comment (dwijnand) [inline proj/lagom.conf]: Cutting 1.6.x (and 2.8.x for Play) tomorrow.

**Comment by SethTisue** [comment]
https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2757/ has

```
[akka-stream] [info]   com.fasterxml.jackson.databind.JsonMappingException: Scala module 2.9.9-dbuildx03dc08a06dfd05001273b7260847a0373a3540d6 requires Jackson Databind version >= 2.9.0 and < 2.10.0
...
[akka-stream] [error] Error during tests:
[akka-stream] [error] 	akka.serialization.jackson.JacksonCborSerializerSpec
[akka-stream] [error] 	akka.serialization.jackson.JacksonJsonSerializerSpec
[akka-stream] [error] 	doc.akka.serialization.jackson.SerializationDocSpec
```

I'll dig it into later

File: proj/lagom.conf (hunk 1 above)
Comment (SethTisue) [reply proj/lagom.conf]: for now, in the playframework repo should I be using 2.7.x, or master?

Comment (dwijnand) [reply proj/lagom.conf]: master for Akka 2.6

**Comment by dwijnand** [comment]
Yeah the Akka/Play/Lagom family also upgraded to Jackson 2.10, which isn't as trivial of a version bump as one might think (or... hope for).

**Comment by dwijnand** [comment]
The branches exist now:

* https://github.com/playframework/playframework/tree/2.8.x
* https://github.com/lagom/lagom/tree/1.6.x

**Comment by SethTisue** [comment]
akka-stream is green now — jackson thing is sorted

* akka-more `object protobufv3 is not a member of package akka
[akka-more] [error] import akka.protobufv3.internal.Any` https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2781/artifact/logs/akka-more-build.log
* akka-http `Detected possible incompatible versions on the classpath. Please note that a given Akka version MUST be the same across all modules of Akka that you are using` https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2781/artifact/logs/akka-http-build.log
* play-ws similar error as akka-http https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2781/artifact/logs/play-ws-build.log

**Comment by SethTisue** [comment]
re: the protobufV3 thing, akka build.sbt has:

```
lazy val protobufV3 = akkaModule("akka-protobuf-v3")
  .settings(AutomaticModuleName.settings("akka.protobuf.v3"))
  .enablePlugins(ScaladocNoVerificationOfDiagrams)
  .disablePlugins(MimaPlugin)
  .settings(
    libraryDependencies += Dependencies.Compile.protobufRuntime,                                                        
    assemblyShadeRules in assembly := Seq(
        ShadeRule
          .rename("com.google.protobuf.**" -> "akka.protobufv3.internal.@1")
          .inLibrary(Dependencies.Compile.protobufRuntime)),
    assemblyOption in assembly := (assemblyOption in assembly).value.copy(includeScala = false, includeBin = false),
    crossPaths := false,
    autoScalaLibrary := false, // do not include scala dependency in pom                                                
    exportJars := true, // in dependent projects, use assembled and shaded jar                                          
    makePomConfiguration := makePomConfiguration.value
        .withConfigurations(Vector(Compile)), // prevent original dependency to be added to pom as runtime dep          
    packageBin in Compile := (assembly in Compile).value, // package by running assembly                                
    // Prevent cyclic task dependencies, see https://github.com/sbt/sbt-assembly/issues/365                             
    fullClasspath in assembly := (managedClasspath in Runtime).value, // otherwise, there's a cyclic dependency between packageBin and assembly                                                                                                
    test in assembly := {}, // assembly runs tests for unknown reason which introduces another cyclic dependency to packageBin via exportedJars                                                                                                
    description := "Akka Protobuf V3 is a shaded version of the protobuf runtime. Original POM: https://github.com/protocolbuffers/protobuf/blob/v3.9.0/java/pom.xml")
```

🙀

**Comment by SethTisue** [comment]
we probably ought to just pull the akka-protobuf-v3 JAR from Maven Central

**Comment by SethTisue** [comment]
`-Dakka.fail-mixed-versions=false` should take care of the mixed version thing. it happens because we build some akka modules in `akka` and some in `akka-stream` (and, though it doesn't enter into the above errors specifically, still others in `akka-more`)

**Comment by dwijnand** [comment]
~In the context of the community build where binary compatibility isn't a concern (except for pulled jars) you're still at risk of source incompatibilities: for example because the older SHA of akka-http uses an internal API that has since been removed.~

Actually, I would have to check because I know internal API usage is mostly an in-repo thing, i.e. akka-http only uses public, stable APIs.

**Comment by SethTisue** [comment]
@dwijnand I have frozen and unfrozen akka/akka-http/et al in tandem or not in tandem, as they drift in and out of sync with each other, many times over the years. now that we used fixed SHAs for everything it has become less of an annoyance/time-sink. but I do still have to deal with it at the time I advance all the SHAs. it generally works fine to just 1) blindly advance the SHAs, 2) usually there's no problem, 3) if it does fail then temporarily freeze something, it's usually not real hard to figure out what needs to be frozen (typically the upstream repo needs to be rewound a bit and frozen, because changes were made that the downstream repo hasn't adapted to yet)

**Comment by SethTisue** [comment]
https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2786/

* protobuf-v3 thing is sorted now
* akka-more
  * `PersistentFsmToTypedMigrationCompileOnlyTest.java:22:1: cannot find symbol`
  * https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2786/artifact/logs/akka-more-build.log
* akka-http and play-ws
  * `-Dakka.fail-mixed-versions=false` seems not to have taken effect somehow?

**Comment by SethTisue** [comment]
progress: akka-http and play-ws are green now; akka-more fixes in the works

**Comment by SethTisue** [comment]
also green now: akka-more, akka-persistence-cassandra. next: playframework

**Comment by SethTisue** [comment]
https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2810/

playframework is green :tada:

I had akka-persistence-cassandra green, but then lagom failed downstream. the obvious fix of just pinning akka-persistence-cassandra to the version lagom expects didn't work:

```
[akka-persistence-cassandra] [error] /home/jenkins/workspace/scala-2.13.x-integrate-community-build/target-0.9.16/project-builds/akka-persistence-cassandra-1d67cdfc4d27f25e075b1630d523cb995a44dc64/core/src/test/scala/akka/persistence/cassandra/journal/CassandraLoadTypedSpec.scala:180:50: overloaded method value apply with alternatives:
[akka-persistence-cassandra] [error]   (entityTypeHint: String,entityId: String,separator: String)akka.persistence.typed.PersistenceId <and>
[akka-persistence-cassandra] [error]   (entityTypeHint: String,entityId: String)akka.persistence.typed.PersistenceId
[akka-persistence-cassandra] [error]  cannot be applied to (String)
[akka-persistence-cassandra] [error]         system.spawnAnonymous(Processor.behavior(PersistenceId("p1"), probe.ref, notifyProbeInEventHandler = false))
[akka-persistence-cassandra] [error]                                                  ^
```

nor did this fix it:

```diff
+  extra.options: ["-Doverride.akka.version=2.6.0"]
```

needs more digging, probably not until next week

the good news is that we are now close.

**Comment by dwijnand** [comment]
That was a late source incompatible change in Akka persistence typed, that it looks isn't in that source.  Maybe it needs a new sha or maybe the test still needs updating.

**Comment by SethTisue** [comment]
this is on hold until the https://github.com/scala/scala/pull/8538 regression is resolved. (working on two different Scala SHAs tends to chew up too much disk space. we could temporarily revert the 2.13.x branch to the old Scala SHA, but meh.)

@dwijnand maybe by the time that's resolved, this PR can become your problem instead of mine :-)

and by the way, a while back you'd suggested we do some pairing on community build work, in order for you to learn it. that now makes even more sense, 1) given your upcoming team change, and 2) because now that the changes listed in https://contributors.scala-lang.org/t/community-build-progress-report-august-2019/3573/8 are complete, the new workflow is something I feel much better about involving others in

**Comment by dwijnand** [comment]
Yep, this looks like a good two-birds-with-one-stone task.  The "`./narrow akka-stream` followed by `./run.sh`" instructions look super-slick!

Let's chat on Tuesday or Wednesday?

**Comment by SethTisue** [comment]
@dwijnand Tuesday is a travel/vacation day for me; send me an invite for Wednesday?

for now, I dealt with the scala/scala#8538 regression by reverting this repo to a known-green state: 7199f890ab9fb53f43278b8d73b6ce3e83aacc2d . so that leaves this PR free to move forward again.

**Comment by SethTisue** [comment]
brought up to date with 2.13.x HEAD

https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2839/

**Comment by dwijnand** [comment]
> Tuesday is a travel/vacation day for me; send me an invite for Wednesday?

Let's chat at+after the team meeting tomorrow.

**Comment by SethTisue** [comment]
https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2841/

**Comment by SethTisue** [comment]
2839 was a botched merge
 
2841 has current status; akka-stream fails, which is a regression since run 2810:

```
[akka-stream] [error] Modules were resolved with conflicting cross-version suffixes in ProjectRef(uri("file:/home/jenkins/workspace/scala-2.13.x-integrate-community-build/target-0.9.16/project-builds/akka-stream-d61b7559285c44fd5a23fca2487c8e2e079019ee/"), "akka-stream"):
[akka-stream] [error]    com.typesafe.akka:akka-protobuf-v3 <none>, _2.13
```

since run 2810, we're on a newer akka SHA, but also a newer sbt version (1.3.3->1.3.4); Dale thinks (in conversation) the 1.3.4 upgrade might actually be relevant

"conflicting cross-version suffix" errors are greatly feared in community build work, it's usually unclear what the cause is, and whether it's dbuild's fault, or what; I usually have to go digging into how I've resolved (or, more usually, flailingly worked around) such errors in the past

**Comment by SethTisue** [comment]
first, let's try downgrading to sbt 1.3.3, just in proj/akka.conf, and see if that un-regresses it:

https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2846/

**Comment by dwijnand** [comment]
Back to that `PersistenceId` source incompatible, late change - in akka-persistence-cassandra; and akka-http needs that "can use akka of different versions" override.

**Comment by SethTisue** [comment]
> akka-http needs that "can use akka of different versions" override

are you sure that's the reason for the test failures? it appears to me — without really digging into it — that the akka version thing is only a warning, but the actual failure is `[akka-http] [ERROR] [12/04/2019 18:50:42.036] [HttpAppSpec-akka.actor.internal-dispatcher-3] [akka://HttpAppSpec/system/IO-TCP/selectors/$a/11] Bind failed for TCP channel on endpoint [localhost/127.0.0.1:39671]`

for now I'll just add an exclude for `HttpAppSpec`

**Comment by SethTisue** [comment]
https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2848/

akka-http is green, so the next hurdle is the akka-persistence-cassandra thing again, like before

**Comment by dwijnand** [comment]
I'll try to understand that one and push something here.

**Comment by SethTisue** [comment]
I see that Akka 2.6.1 came out, so I did a fresh round of SHA advancing

let's see how Jenkins likes it: https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2862/ queued

**Comment by dwijnand** [comment]
Looks to me like you tried rerunning it and a slightly different spec (but still in the same file) failed.  So my exclusion handles that too.

https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2865/

**Comment by SethTisue** [comment]
```
[akka-more] [error] Failed tests:
[akka-more] [error] 	akka.actor.typed.AskSpec
```

queued: https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2867/

**Comment by dwijnand** [comment]
OK, didn't flake again.  Now it's back to the source incompat:

```
[akka-persistence-cassandra] [error] /home/jenkins/workspace/scala-2.13.x-integrate-community-build/target-0.9.16/project-builds/akka-persistence-cassandra-95ea095bf4ba7f0172440861a3696431427b758e/core/src/test/scala/akka/persistence/cassandra/journal/CassandraLoadTypedSpec.scala:180:50: overloaded method value apply with alternatives:
[akka-persistence-cassandra] [error]   (entityTypeHint: String,entityId: String,separator: String)akka.persistence.typed.PersistenceId <and>
[akka-persistence-cassandra] [error]   (entityTypeHint: String,entityId: String)akka.persistence.typed.PersistenceId
[akka-persistence-cassandra] [error]  cannot be applied to (String)
[akka-persistence-cassandra] [error]         system.spawnAnonymous(Processor.behavior(PersistenceId("p1"), probe.ref, notifyProbeInEventHandler = false))
[akka-persistence-cassandra] [error]                                                  ^
[akka-persistence-cassandra] [error] /home/jenkins/workspace/scala-2.13.x-integrate-community-build/target-0.9.16/project-builds/akka-persistence-cassandra-95ea095bf4ba7f0172440861a3696431427b758e/core/src/test/scala/akka/persistence/cassandra/journal/CassandraLoadTypedSpec.scala:189:50: overloaded method value apply with alternatives:
[akka-persistence-cassandra] [error]   (entityTypeHint: String,entityId: String,separator: String)akka.persistence.typed.PersistenceId <and>
[akka-persistence-cassandra] [error]   (entityTypeHint: String,entityId: String)akka.persistence.typed.PersistenceId
[akka-persistence-cassandra] [error]  cannot be applied to (String)
[akka-persistence-cassandra] [error]         system.spawnAnonymous(Processor.behavior(PersistenceId("p2"), probe.ref, notifyProbeInEventHandler = true))
[akka-persistence-cassandra] [error]                                                  ^
[akka-persistence-cassandra] [error] two errors found
[akka-persistence-cassandra] [error] (core / Test / compileIncremental) Compilation failed
```

**Comment by dwijnand** [comment]
oh... https://github.com/akka/akka-persistence-cassandra/blob/master/core/src/test/scala/akka/persistence/cassandra/journal/CassandraLoadTypedSpec.scala#L5

> `/* FIXME enable again when branch for Akka 2.6.x`

**Comment by dwijnand** [comment]
https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2869/ 404 until Alpha gets bored of sleeping on my lap.

**Comment by dwijnand** [comment]
That bombed from FS issues?  Seth triggered https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2872/

Looks like we're past APC, bombing in Lagom with:

```
[lagom] [error] Modules were resolved with conflicting cross-version suffixes in ProjectRef(uri("file:/home/jenkins/workspace/scala-2.13.x-integrate-community-build/target-0.9.16/project-builds/lagom-307c3e97f5f00d512997274e9484ff770ce5ac04/"), "persistence-jdbc-core"):
[lagom] [error]    com.typesafe.akka:akka-persistence-query _2.13, <none>
[lagom] [error]    org.scala-lang.modules:scala-collection-compat <none>, _2.13
[lagom] [error]    com.typesafe.akka:akka-actor <none>, _2.13
[lagom] [error]    com.typesafe.akka:akka-persistence _2.13, <none>
[lagom] [error]    com.typesafe.akka:akka-stream _2.13, <none>
[lagom] [error]    org.scala-lang.modules:scala-java8-compat _2.13, <none>
[lagom] [error]    org.scala-lang.modules:scala-parser-combinators <none>, _2.13
[lagom] [error]    com.typesafe:ssl-config-core <none>, _2.13
```

"conflicting cross-version suffixes" 😨

**Comment by SethTisue** [comment]
I think the conflicting-cross-version-suffixes thing is not the real error, the real error (which I failed to spot before) is `the library com.github.dnvriend#akka-persistence-jdbc is not provided`, and presumably the fix you suggested at #1016 will work here as well. I'll give it a try

note that dbuild kept on trucking, despite the missing dependency, because `proj/lagom.conf` has `check-missing: false` because I don't know any other way to work around the missing dependency error the presence of the sbt scripted-plugin produces

**Comment by SethTisue** [comment]
queued: https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2876/

**Comment by SethTisue** [comment]
queued: https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2876/

**Comment by SethTisue** [comment]
uh, I think I went too fast on that akka-persistence-jdbc thing, and wasn't really paying attention, hoping it was an easy fix. digging...

**Comment by dwijnand** [comment]
https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2878/ is the next run (when it starts)

**Comment by SethTisue** [comment]
the conflicting-cross-version-suffixes thing is not the real error, the real error (which I failed to spot before) is `the library com.github.dnvriend#akka-persistence-jdbc is not provided`

note that dbuild kept on trucking, despite the missing dependency, because `proj/lagom.conf` has `check-missing: false` because I don't know any other way to work around the missing dependency error the presence of the sbt scripted-plugin produces

anyway, I think I have now found an akka-persistence-jdbc SHA that is new enough to work but old enough that it doesn't include the org change 🤞

**Comment by SethTisue** [comment]
https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2896/

**Comment by SethTisue** [comment]
the eagle has landed

fyi @marcospereira @patriknw

**Pull request merged**